pkg/monitor/agent/agent.go

```diff
@@ -69,34 +69,55 @@ type Agent struct {
 }
 
 // NewAgent starts a new monitor agent instance which distributes monitor events
-// to registered listeners. It spawns a singleton goroutine reading events from
+// to registered listeners. Once the datapath is set up, AttachToEventsMap needs
+// to be called to receive events from the perf ring buffer. Otherwise, only
+// user space events received via SendEvent are distributed registered listeners.
+// Internally, the agent spawns a singleton goroutine reading events from
 // the BPF perf ring buffer and provides an interface to pass in non-BPF events.
 // The instance can be stopped by cancelling ctx, which will stop the perf reader
 // go routine and close all registered listeners.
 // Note that the perf buffer reader is started only when listeners are
 // connected.
-func NewAgent(ctx context.Context, nPages int) (a *Agent, err error) {
+func NewAgent(ctx context.Context) *Agent {
+	return &Agent{
+		ctx:              ctx,
+		listeners:        make(map[listener.MonitorListener]struct{}),
+		consumers:        make(map[consumer.MonitorConsumer]struct{}),
+		perfReaderCancel: func() {}, // no-op to avoid doing null checks everywhere
+	}
+}
+
+// AttachToEventsMap opens the events perf ring buffer and makes it ready for
+// consumption, such that any subscribed consumers may receive events
+// from it. This function is to be called once the events map has been set up.
+func (a *Agent) AttachToEventsMap(nPages int) error {
 	// assert that we can actually connect the monitor
 	path := oldBPF.MapPath(eventsMapName)
 	eventsMap, err := ebpf.LoadPinnedMap(path, nil)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	a = &Agent{
-		ctx:              ctx,
-		listeners:        make(map[listener.MonitorListener]struct{}),
-		consumers:        make(map[consumer.MonitorConsumer]struct{}),
-		perfReaderCancel: func() {}, // no-op to avoid doing null checks everywhere
-		events:           eventsMap,
-		MonitorStatus: models.MonitorStatus{
-			Cpus:     int64(eventsMap.MaxEntries()),
-			Npages:   int64(nPages),
-			Pagesize: int64(os.Getpagesize()),
-		},
+	a.Lock()
+	defer a.Unlock()
+
+	if a.events != nil {
```

Comment:
shouldn't this check be before we call `LoadPinnedMap`?

Reply:
Yes, we should. Especially since the code doesn't close the map if this case is hit in the current code. Will fix.

Reply:
Moved the call `LoadPinnedMap` below the check. Unfortunately this means that we need to keep the lock while loading the map (to avoid a TOCTOW issue with `a.events`), but I think that should be fine.